File: src/src/main/resources/mmlspark/ComputePerInstanceStatistics.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class ComputePerInstanceStatistics(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """
    Evaluates the given scored dataset with per instance metrics.

    The Regression metrics are:

    - "L1_loss"
    - "L2_loss"

    The Classification metrics are:

    - "log_loss"

    Args:

        evaluationMetric (str): Metric to evaluate models with (default: all)
        labelCol (str): The name of the label column
        scoredLabelsCol (str): Scored labels column name, only required if using SparkML estimators
        scoredProbabilitiesCol (str): Scored probabilities, usually calibrated from raw scores, only required if using SparkML estimators
        scoresCol (str): Scores or raw prediction column name, only required if using SparkML estimators
    """

    @keyword_only
    def __init__(self, evaluationMetric="all", labelCol=None, scoredLabelsCol=None, scoredProbabilitiesCol=None, scoresCol=None):
        super(ComputePerInstanceStatistics, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.ComputePerInstanceStatistics")
        self.evaluationMetric = Param(self, "evaluationMetric", "evaluationMetric: Metric to evaluate models with (default: all)")
        self._setDefault(evaluationMetric="all")
        self.labelCol = Param(self, "labelCol", "labelCol: The name of the label column")
        self.scoredLabelsCol = Param(self, "scoredLabelsCol", "scoredLabelsCol: Scored labels column name, only required if using SparkML estimators")
        self.scoredProbabilitiesCol = Param(self, "scoredProbabilitiesCol", "scoredProbabilitiesCol: Scored probabilities, usually calibrated from raw scores, only required if using SparkML estimators")
        self.scoresCol = Param(self, "scoresCol", "scoresCol: Scores or raw prediction column name, only required if using SparkML estimators")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, evaluationMetric="all", labelCol=None, scoredLabelsCol=None, scoredProbabilitiesCol=None, scoresCol=None):
        """
        Set the (keyword only) parameters

        Args:

            evaluationMetric (str): Metric to evaluate models with (default: all)
            labelCol (str): The name of the label column
            scoredLabelsCol (str): Scored labels column name, only required if using SparkML estimators
            scoredProbabilitiesCol (str): Scored probabilities, usually calibrated from raw scores, only required if using SparkML estimators
            scoresCol (str): Scores or raw prediction column name, only required if using SparkML estimators
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setEvaluationMetric(self, value):
        """

        Args:

            evaluationMetric (str): Metric to evaluate models with (default: all)

        """
        self._set(evaluationMetric=value)
        return self


    def getEvaluationMetric(self):
        """

        Returns:

            str: Metric to evaluate models with (default: all)
        """
        return self.getOrDefault(self.evaluationMetric)


    def setLabelCol(self, value):
        """

        Args:

            labelCol (str): The name of the label column

        """
        self._set(labelCol=value)
        return self


    def getLabelCol(self):
        """

        Returns:

            str: The name of the label column
        """
        return self.getOrDefault(self.labelCol)


    def setScoredLabelsCol(self, value):
        """

        Args:

            scoredLabelsCol (str): Scored labels column name, only required if using SparkML estimators

        """
        self._set(scoredLabelsCol=value)
        return self


    def getScoredLabelsCol(self):
        """

        Returns:

            str: Scored labels column name, only required if using SparkML estimators
        """
        return self.getOrDefault(self.scoredLabelsCol)


    def setScoredProbabilitiesCol(self, value):
        """

        Args:

            scoredProbabilitiesCol (str): Scored probabilities, usually calibrated from raw scores, only required if using SparkML estimators

        """
        self._set(scoredProbabilitiesCol=value)
        return self


    def getScoredProbabilitiesCol(self):
        """

        Returns:

            str: Scored probabilities, usually calibrated from raw scores, only required if using SparkML estimators
        """
        return self.getOrDefault(self.scoredProbabilitiesCol)


    def setScoresCol(self, value):
        """

        Args:

            scoresCol (str): Scores or raw prediction column name, only required if using SparkML estimators

        """
        self._set(scoresCol=value)
        return self


    def getScoresCol(self):
        """

        Returns:

            str: Scores or raw prediction column name, only required if using SparkML estimators
        """
        return self.getOrDefault(self.scoresCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.ComputePerInstanceStatistics"

    @staticmethod
    def _from_java(java_stage):
        module_name=ComputePerInstanceStatistics.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".ComputePerInstanceStatistics"
        return from_java(java_stage, module_name)

File: src/main/scala/com/microsoft/ml/spark/train/ComputePerInstanceStatistics.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.train

import com.microsoft.ml.spark.core.contracts._
import com.microsoft.ml.spark.core.metrics.{MetricConstants, MetricUtils}
import com.microsoft.ml.spark.core.schema.{CategoricalUtilities, SchemaConstants, SparkSchema}
import com.microsoft.ml.spark.core.schema.SchemaConstants._
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util.{DefaultParamsReadable, DefaultParamsWritable, Identifiable}
import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.types.injections.MetadataUtilities

object ComputePerInstanceStatistics extends DefaultParamsReadable[ComputePerInstanceStatistics] {
  val epsilon = 1e-15
}

trait ComputePerInstanceStatisticsParams extends Wrappable with DefaultParamsWritable
  with HasLabelCol with HasScoresCol with HasScoredLabelsCol with HasScoredProbabilitiesCol with HasEvaluationMetric {
  /** Param "evaluationMetric" is the metric to evaluate the models with. Default is "all"
    *
    *   If using a native Spark ML model, you will need to specify either "classifier" or "regressor"
    *     - classifier
    *     - regressor
    *
    * @group param
    */
  setDefault(evaluationMetric -> MetricConstants.AllSparkMetrics)
}

/** Evaluates the given scored dataset with per instance metrics.
  *
  * The Regression metrics are:
  * - L1_loss
  * - L2_loss
  *
  * The Classification metrics are:
  * - log_loss
  */
class ComputePerInstanceStatistics(override val uid: String) extends Transformer
  with ComputePerInstanceStatisticsParams {

  def this() = this(Identifiable.randomUID("ComputePerInstanceStatistics"))

  override def transform(dataset: Dataset[_]): DataFrame = {
    val (modelName, labelColumnName, scoreValueKind) =
      MetricUtils.getSchemaInfo(
        dataset.schema,
        if (isDefined(labelCol)) Some(getLabelCol) else None,
        getEvaluationMetric)

    val dataframe = dataset.toDF()

    if (scoreValueKind == SchemaConstants.ClassificationKind) {
      // Compute the LogLoss for classification case
      val scoredLabelsColumnName =
        if (isDefined(scoredLabelsCol)) getScoredLabelsCol
        else SparkSchema.getScoredLabelsColumnName(dataframe, modelName)

      // Get levels if categorical
      val levels = CategoricalUtilities.getLevels(dataframe.schema, labelColumnName)
      val numLevels =
        if (!levels.isEmpty && levels.get != null) {
          if (levels.get.length > 2) levels.get.length else 2
        } else {
          // Otherwise compute unique levels
          dataset.select(col(labelColumnName).cast(DoubleType)).rdd.distinct().count().toInt
        }

      val logLossFunc = udf((scoredLabel: Double, scores: org.apache.spark.ml.linalg.Vector) =>
        if (scoredLabel < numLevels) {
          -Math.log(Math.min(1, Math.max(ComputePerInstanceStatistics.epsilon, scores(scoredLabel.toInt))))
        } else {
          // penalize if no label seen in training
          -Math.log(ComputePerInstanceStatistics.epsilon)
        })
      val probabilitiesColumnName =
        if (isDefined(scoredProbabilitiesCol)) getScoredProbabilitiesCol
        else SparkSchema.getScoredProbabilitiesColumnName(dataframe, modelName)
      dataframe.withColumn(MetricConstants.LogLossMetric,
        logLossFunc(dataset(scoredLabelsColumnName), dataset(probabilitiesColumnName)))
    } else {
      val scoresColumnName =
        if (isDefined(scoresCol)) getScoresCol
        else SparkSchema.getScoresColumnName(dataframe, modelName)
      // Compute the L1 and L2 loss for regression case
      val scoresAndLabels =
        dataset.select(col(scoresColumnName), col(labelColumnName).cast(DoubleType)).rdd.map {
          case Row(prediction: Double, label: Double) => (prediction, label)
        }
      val l1LossFunc = udf((trueLabel:Double, scoredLabel: Double) => math.abs(trueLabel - scoredLabel))
      val l2LossFunc = udf((trueLabel:Double, scoredLabel: Double) =>
        {
          val loss = math.abs(trueLabel - scoredLabel)
          loss * loss
        })
      dataframe.withColumn(MetricConstants.L1LossMetric,
        l1LossFunc(dataset(labelColumnName), dataset(scoresColumnName)))
        .withColumn(MetricConstants.L2LossMetric,
          l2LossFunc(dataset(labelColumnName), dataset(scoresColumnName)))
    }
  }

  private def getFirstModelName(colMetadata: Metadata): Option[String] = {
    if (!colMetadata.contains(MMLTag)) null
    else {
      val mlTagMetadata = colMetadata.getMetadata(MMLTag)
      val metadataKeys = MetadataUtilities.getMetadataKeys(mlTagMetadata)
      metadataKeys.find(key => key.startsWith(SchemaConstants.ScoreModelPrefix))
    }
  }

  override def copy(extra: ParamMap): Transformer = new ComputePerInstanceStatistics()

  // TODO: This should be based on the retrieved score value kind
  override def transformSchema(schema: StructType): StructType =
    schema.add(new StructField(MetricConstants.L1LossMetric, DoubleType))
          .add(new StructField(MetricConstants.L2LossMetric, DoubleType))

}
